TASK: Simple Nullable Handling

mhsdesign opened the pull request:
- Scopes will check if $typeReferenceNode->isOptional and if so, wrap the type in a union with null
- The TypeReferenceTranspiler can now transpile those simple unions, by looking into its members

mhsdesign commented on src/TypeSystem/Type/UnionType/UnionType.php:
```diff
@@ -53,7 +60,32 @@ public static function of(TypeInterface ...$members): TypeInterface
+            }
+            $nonNullMembers[] = $member;
+        }
+        if (count($nonNullMembers) === 1) {
```
#premature performance optimization is the root of all evil

mhsdesign commented on src/TypeSystem/Type/UnionType/UnionType.php:
```diff
@@ -22,17 +22,24 @@
 
     private function __construct(TypeInterface ...$members)
     {
+        if (count($members) < 1) {
```
Wow thanks, great idea with enforcing it structurally `TypeInterface $firstMember, TypeInterface ...$members` didnt think about this ^^

i kept the 

```
assert(count($members) > 1, 'UnionType must hold at least two different members');
```

in the constructor, since i like having the rules of objects enforced in the constructor (in case someone creates a new static factory or a `with*` method, which invokes the private constructor directly ...)

mhsdesign commented on src/TypeSystem/Resolver/TernaryOperation/TernaryOperationTypeResolver.php:
```diff
@@ -41,14 +44,39 @@ public function resolveTypeOf(TernaryOperationNode $ternaryOperationNode): TypeI
+                $falseExpressionTypeResolver->resolveTypeOf($ternaryOperationNode->false)
+            );
+        }
+
```
Implemented :D But now im too tired to type things down that i encountered, see you tomorrow.

(And thanks for the tipp with TernaryBranchScope works like a charm and even better ^^)

Also i cant thank you enough for your most precise code reviews and comments - they are extremely helpful and well guiding me! Thank you really so much!!!

mhsdesign replied:
yes youre right, my implementation is only limited for super simple things like `foo ? foo : ""`

i thought about how we can i implement a universal solution, which also handles any thinkable more complex case, but it seems like a hard challenge

for example given `foo` is a nullable string.

those expressions will be resolved as of now correctly:

```
foo ? foo : ""
foo !== null ? foo : ""
```

but here we loose it:

```
(foo) ? foo : ""
foo !== nullVariable ? foo : ""
((foo)) ? foo : ""
(foo !== null) ? foo : ""
foo && true ? foo : ""
(foo ? foo : foo) ? foo : ""
((foo ? foo : foo) ? foo : foo) ? foo : ""
```

I checked all the examples with typescript, and it does a good job (unsurprisingly). I even went as far as trying to have a look at the ts codebase for some inspiration, but i couldnt even find the right unit tests to start with xD

Some of the above are surely not a real world use case but its a simple demonstration how easy it is to confuse the implemented type resolution (or type inference?).

I think i need to find out while resolving the condition part, which variable participated and how it affected the truthyness. Im not really sure if this would be the right approach in the long run and if it can be easily implemented.

But for now im also quite happy with my current implementation as a start.

mhsdesign replied:
Found the place where phpstan does it:
https://github.com/phpstan/phpstan-src/blob/07bb4aa2d5e39dafa78f56c5df132c763c2d1b67/src/Analyser/MutatingScope.php#L1616

```php
// $node is a TernaryExpression with the fields: cond, if and else

$booleanConditionType = $this->getType($node->cond)->toBoolean();
if ($booleanConditionType->isTrue()->yes()) {
     return $this->filterByTruthyValue($node->cond)->getType($node->if);
}

if ($booleanConditionType->isFalse()->yes()) {
    return $this->filterByFalseyValue($node->cond)->getType($node->else);
}

return TypeCombinator::union(
     $this->filterByTruthyValue($node->cond)->getType($node->if),
     $this->filterByFalseyValue($node->cond)->getType($node->else),
);
```

grebaldi commented on src/TypeSystem/Inferrer/InferredTypes.php:
```diff
@@ -0,0 +1,72 @@
+        return $this->types[$identifierName] ?? null;
+    }
+
+    /**
+     * @template T
+     * @param array<string|int,T> $array
+     * @phpstan-assert-if-true array<string,T> $array
+     */
+    private static function isAssociativeArray(array $array): bool
+    {
+        foreach ($array as $key => $value) {
+            if (is_string($key)) {
+                continue;
+            }
+            return false;
+        }
+        return true;
+    }
```
```suggestion
```

grebaldi commented on src/TypeSystem/Inferrer/InferredTypes.php:
```diff
@@ -0,0 +1,72 @@
+
+use PackageFactory\ComponentEngine\TypeSystem\TypeInterface;
+
+class InferredTypes
```
```suggestion
class TypeMap
```

It might be sensible to move this class to a namespace `PackageFactory\ComponentEngine\TypeSystem\Utility` or so. It may have more uses than just this one.

mhsdesign replied:
Still not sure ;) maybe another day ^^ It will fall all into place sometime...

grebaldi commented on src/TypeSystem/Inferrer/InferredTypes.php:
```diff
@@ -0,0 +1,72 @@
+    private function __construct(
+        TypeInterface ...$types
+    ) {
+        assert(self::isAssociativeArray($types), '$types must be an associative array');
```
```suggestion
```

It's fair to assume that it is. The class itself is in control of this (via static factory methods).

I know, it's not perfect, since anyone who changes this class must be aware of this implicit assertion. Yet, on the other hand: What can really go wrong here? If `$types` has numeric keys, they will be ignored - nothing breaks.

mhsdesign replied:
Yes was just trying out other things :D It would be cool if you could annotate and tell phpstan, that the the varadic/spread should be with string keys. So it is statically enforced.

grebaldi commented on src/TypeSystem/Inferrer/InferredTypes.php:
```diff
@@ -0,0 +1,72 @@
+        return new self();
+    }
+
+    public static function fromType(string $identifierName, TypeInterface $type): self
```
```suggestion
    public static function fromEntry(string $identifierName, TypeInterface $type): self
```

mhsdesign replied:
Nice one ☝️ like that naming convention!

grebaldi commented on src/TypeSystem/Inferrer/TypeInferrer.php:
```diff
@@ -0,0 +1,97 @@
+
+declare(strict_types=1);
+
+namespace PackageFactory\ComponentEngine\TypeSystem\Inferrer;
```
```suggestion
namespace PackageFactory\ComponentEngine\TypeSystem\Narrower\Expression;
```

Following the `TypeSystem\[Functionality]\[AstNodeType]` pattern.

grebaldi commented on src/TypeSystem/Inferrer/TypeInferrer.php:
```diff
@@ -0,0 +1,97 @@
+ * The structure is partially inspired by phpstan
+ * https://github.com/phpstan/phpstan-src/blob/07bb4aa2d5e39dafa78f56c5df132c763c2d1b67/src/Analyser/TypeSpecifier.php#L111
+ */
+class TypeInferrer
```
```suggestion
class ExpressionTypeNarrower
```

grebaldi commented on src/TypeSystem/Inferrer/TypeInferrer.php:
```diff
@@ -0,0 +1,97 @@
+    ) {
+    }
+
+    public function inferTypesInCondition(ExpressionNode $conditionNode, TypeInferrerContext $context): InferredTypes
```
```suggestion
    public function narrowTypesOfSymbolsIn(ExpressionNode $expressionNode): TypeMap
```

Imho, we do not need to constraint this particular algorithm to ternary condition nodes. Type narrowing can be done given any expression.

Furthermore, as far as I understand, `TypeInferrerContext` is supposed to inform this method of whether we assume the ternary condition to be truthy or falsy. I don't believe we need to make this distinction. It is sufficient to perform the narrowing algorithm under the assumption that `$expressionNode` evaluates to be truthy. Given the result of that, for the assumption that `$expressionNode` evaluates to be falsy the opposite must be true (which means that `UnionType` may need another method to build the difference with a given `UnionType`).

For performance it may be wise to have this class cache the result of this method in memory and let both `TernaryBranchScope`s share one instance.

mhsdesign replied:
> Furthermore, as far as I understand, TypeInferrerContext is supposed to inform this method of whether we assume the ternary condition to be truthy or falsy.

Exactly ;)

>  I don't believe we need to make this distinction. It is sufficient to perform the narrowing algorithm under the assumption that $expressionNode evaluates to be truthy. Given the result of that, for the assumption that $expressionNode evaluates to be falsy the opposite must be true (which means that UnionType may need another method to build the difference with a given UnionType).

Hmm i was convinced by this approach at first but i have the feeling that important informations will be lost when narrowing the true branch, and its not (with out current TypeMap) not possible to invert it correctly:

`foo ? true : false`

if `foo` is of type `string|null` the true ast will know its `string` and inverting is not possible since we dont know yet if this was ever a nullable type - we cant just make it null for the false branch, as it might not have been nullable in the first place. Also the idea was copied from phpstan and in my head it makes perfectly sense when the expression contains complex logic, but i cant bring up another example right now ^^

Either way id like to propose to stick to it for know and refactor later if we dont need it or find it "too slow".

grebaldi replied:
You are very right! I had a thinking error there :sweat_smile: 

My statement
> Given the result of [the narrowing algorithm under the assumption that $expressionNode evaluates to be truthy], for the assumption that $expressionNode evaluates to be falsy the opposite must be true

can easily be falsified:

```typescript
foo: null|string
bar: null|string

foo && bar
  ? (/* foo and bar MUST be of type string */)
  : (/* foo and bar MAY each be of type null|string */)
```

So, it is inevitable to make the distinction. However, I'd still like to suggest two changes here:

1. Rename `TypeInferrerContext` to `Truthiness`
2. Move the parameter that is currently called `$context` to the constructor (and probably rename it to something like `$assumedTruthiness` - not sure right now whether that's precise enough), so the resulting method `narrowTypesOfSymbolsIn` becomes unary

Wdyt?

mhsdesign replied:
Adjusted ;)

Discussed using instance property for `$assumedTruthiness` with @fcool and he found it also a good idea, so we did it ^^

grebaldi commented on src/TypeSystem/Inferrer/TypeInferrer.php:
```diff
@@ -0,0 +1,97 @@
+            if (!$type) {
+                return InferredTypes::empty();
+            }
+            // case `nullableString ? "nullableString is not null" : "nullableString is null"`
+            return InferredTypes::fromType($conditionNode->root->value, $context->narrowDownType($type));
+        }
+
+        if (($binaryOperationNode = $conditionNode->root) instanceof BinaryOperationNode) {
+            // cases
+            // `nullableString === null ? "nullableString is null" : "nullableString is not null"`
+            // `nullableString !== null ? "nullableString is not null" : "nullableString is null"`
+            if (count($binaryOperationNode->operands->rest) !== 1) {
+                return InferredTypes::empty();
+            }
+            $first = $binaryOperationNode->operands->first;
+            $second = $binaryOperationNode->operands->rest[0];
+
+            $comparedIdentifierValueToNull = match (true) {
+                // case `nullableString === null`
+                $first->root instanceof IdentifierNode && $second->root instanceof NullLiteralNode => $first->root->value,
+                // yodas case `null === nullableString`
+                $first->root instanceof NullLiteralNode && $second->root instanceof IdentifierNode => $second->root->value,
+                default => null
+            };
+
+            if ($comparedIdentifierValueToNull === null) {
+                return InferredTypes::empty();
+            }
+            $type = $this->scope->lookupTypeFor($comparedIdentifierValueToNull);
+            if (!$type) {
+                return InferredTypes::empty();
+            }
+
+            if ($binaryOperationNode->operator === BinaryOperator::EQUAL) {
+                return InferredTypes::fromType($comparedIdentifierValueToNull, $context->negate()->narrowDownType($type));
+            }
+            if ($binaryOperationNode->operator === BinaryOperator::NOT_EQUAL) {
+                return InferredTypes::fromType($comparedIdentifierValueToNull, $context->narrowDownType($type));
+            }
+        }
```
These cases may be split into separate `*Narrower` classes - similarly to how the `*Resolver`-classes are organized. But I think it's okay to leave them in this class for now.

mhsdesign replied:
I will wait till this class is more complex to see what other usecase we also have to solve and then later split it up.

grebaldi commented on src/TypeSystem/Inferrer/TypeInferrerContext.php:
See comments for `TypeInferrer`: I believe this class can be removed :)

grebaldi commented on src/TypeSystem/Scope/TernaryBranchScope/TernaryBranchScope.php:
```diff
@@ -0,0 +1,66 @@
+    ) {
+    }
+
+    public static function forTruthyBranch(ExpressionNode $conditionNode, ScopeInterface $parentScope): self
+    {
+        return new self(
+            (new TypeInferrer($parentScope))->inferTypesInCondition($conditionNode, TypeInferrerContext::TRUTHY),
+            $parentScope
+        );
+    }
+
+    public static function forFalsyBranch(ExpressionNode $conditionNode, ScopeInterface $parentScope): self
+    {
+        return new self(
+            (new TypeInferrer($parentScope))->inferTypesInCondition($conditionNode, TypeInferrerContext::FALSY),
+            $parentScope
+        );
+    }
```
As mentioned in the comments for `TypeInferrer.php`, the `TernaryOperationTypeResolver` class should take care of instantiating one `TypeNarrower` that both `TernaryBranchScope`s can share.

This has consequences for the signature of these two methods and...

grebaldi replied:
nvm. (see: https://github.com/PackageFactory/PackageFactory.ComponentEngine/pull/14#discussion_r1179451334)

grebaldi commented on src/TypeSystem/Scope/TernaryBranchScope/TernaryBranchScope.php:
```diff
@@ -0,0 +1,66 @@
+        );
+    }
+
+    public function lookupTypeFor(string $name): ?TypeInterface
+    {
+        return $this->inferredTypes->getType($name) ?? $this->parentScope->lookupTypeFor($name);
+    }
+
```
...the body of this one. `TernaryBranchScope` now needs to be aware of whether it corresponds to the `true`- or the `false`-branch (a boolean flag for this is sufficient). In the `false` case it needs to interpret the result from the narrower differently.

grebaldi replied:
nvm. (see: https://github.com/PackageFactory/PackageFactory.ComponentEngine/pull/14#discussion_r1179451334)

mhsdesign commented on src/TypeSystem/Type/UnionType/UnionType.php:
```diff
@@ -67,15 +94,26 @@ public function is(TypeInterface $other): bool
         }
     }
+
+    /** @return \ArrayIterator<int, TypeInterface> */
+    public function getIterator(): \ArrayIterator
+    {
+        return new \ArrayIterator($this->members);
+    }
```
Ha no thanks, just from muscle memory ^^ didn’t thought about the other way ^^

mhsdesign replied:
I usually use array iterator, so I can get the ArrayCopy in edgecases ^^

But for that one should better implement toArray ^^

mhsdesign commented:
Hi @grebaldi ;)

No problem at all. Im happy for your feedback :D

Before addressing your comments id like to merge https://github.com/PackageFactory/PackageFactory.ComponentEngine/pull/16 and https://github.com/PackageFactory/PackageFactory.ComponentEngine/pull/19 as they both will affect this pr as well.
With those changes i want to try implementing a narrower for:

`!foo ? 1 : 0` and `!!foo ? 1 : 0`
and
`foo === null === true ? 1 : 0`

that way i can see if my the approach is flexible enough to allow even more advanced narrowing.

Later my goal is also having a narrower for `foo.bar ? 1 : 0`... lets see.

grebaldi commented on src/TypeSystem/Scope/TernaryBranchScope/TernaryBranchScope.php:
```diff
@@ -26,13 +26,12 @@
 final class TernaryBranchScope implements ScopeInterface
 {
-    private function __construct(
+    public function __construct(
```
Why that?

mhsdesign replied:
Because one should use the static factories - but I guess this will make unit testing harder … for now fine imo? I like having private constructors at times …

grebaldi replied:
But you've made it `public` :monocle_face:

mhsdesign replied:
Misread the diff 😂 its the 4 day in berlin 😅😂

mhsdesign commented:
Btw, as discussed TypeChecking is for now mostly a tool for helping with the transpilation. You said that at the current stage we dont want and need 100% accuracy. So for the better of the project it might make sense to not merge this for now (even if i am happy with the code) because it increases the complexity quite a bit (adding the new TypeNarrower concept).

It was definitely fun implementing it so no bad blood ;)

I will extract soonish the non TypeNarrower related concepts into another PR and leave this WIP.
